data_analysis/data_capture.py

```diff
@@ -16,25 +16,31 @@
 
 from data_types import Recording, RecordingEnvironment, Event
 
-
 # Name of your serial port (e.g., '/dev/cu.usbserial-0001', or COM5).
-def collect_data(seconds: float, serial_port = '/dev/cu.usbserial-0001'):
+def collect_data(seconds: float, serial_port = '/dev/cu.usbserial-0001', fs=None, error_threshold=0.5):
     # Open the serial port for communication with the ESP8266.
     ser = serial.Serial(serial_port, 115200)
     t_end = time.time() + seconds # Data collection will run for this many seconds
     vibes = [] # Raw accelerometer data
     button_presses = [] # Timestamps of button presses
     start_time = None
+    last_time = None
 
     try:
         while time.time() < t_end:
             timestamp, accel, event = ser.readline().decode().strip().split(',') # Read data from the serial port.
+            timestamp = int(timestamp)/1000
+            if last_time is not None and fs is not None:
+                fs_error = (timestamp - last_time) - (1/fs)
+                if (fs_error) >= (1/fs)*error_threshold:
+                    raise EnvironmentError(f'Expected {1/fs} sampling period, received {timestamp - last_time}')
+            last_time = timestamp
             print(accel) # Print the data to the console.
             vibes.append(float(accel))
             if start_time is None:
                 start_time = timestamp
-            if event is not None:
-              button_presses.append(timestamp - start_time)
+            if event == "BUTTON":
+                button_presses.append(timestamp - start_time)
     except KeyboardInterrupt:
         # Close the serial port and the file when you interrupt the script.
         ser.close()
```

Comment:
I'm not entirely sure this try-except block needs to exist. Try running without it.

If it is needed, let's switch it to a context manager instead (but that doesn't have to happen in this PR)

Reply:
goal was to give an option for us to stop data collection if we're unsure how long we want to collect for. when I was doing initial collection I was like dammit it was too long or not too long and had to start over.